### DISCUSSION TITLE
Support of Array with multiple elements (such as PxrRamp)

### DISCUSSION THREAD

**Discussion by BastienLaby** [Autodesk/maya-usd#1040] [Q&A]
Hi everyone, thanks for all your work here.

I would like to export a shader with a PxrRamp ont it, but exporting arrays with multiple elements seems not supported since the Pixar version of the plugin.

To reproduce, create a basic sphere with a material including a PxrRamp (one example scene attached, using Renderman 22.6). Export the sphere with the following MEL command :
`file -force -options "shadingMode=pxrRis;" -typ "pxrUsdExport" -pr -es "E:/pxrramp.usda";`

Result :
```
Warning: Array with multiple elements encountered at 'PxrRamp1.colorRamp'. Currently, only arrays with a single element are supported. // 
```

Does a workaround exists ?

Thanks

**Comment by mattyjams** [comment]
Hi @BastienLaby,

Yeah, unfortunately that's correct. The old (and now deprecated) `pxrRis` shadingMode had a restriction that only single-element array attributes were supported for export. With the new `useRegistry` shadingMode and the `rendermanForMaya` material conversion setting, you should get the same resulting USD when exporting RfM shader networks:

`... -options "shadingMode=useRegistry;convertMaterialsTo=rendermanForMaya;" ...`

But since the new "generic" shader writer was adapted from the original pxrRis one, the array attribute limitation currently remains. It hopefully shouldn't be a major project to add that support, but no one has had a chance to attack it just yet.

There isn't currently a workaround available in the shipping plugins, but you could try adding the support yourself, or writing your own shader writer. When you use the `useRegistry`/`rendermanForMaya` options I mentioned above, this would be the area of the code you'd want to look at:

https://github.com/Autodesk/maya-usd/blob/d379bfcef63b61036d4b067dbf234be572d4239c/lib/mayaUsd/fileio/shading/symmetricShaderWriter.cpp#L130

Some of the other functions that come after that may or may not do the right thing when presented with array attributes, so making sure they're handled correctly through all of that would likely be the bulk of the work.

So I'm not sure if that helps, but that's the current state of things. Let us know if you have any other questions, or if this becomes a major blocker for you and maybe someone can free up some time to take a crack at addressing this.

Thanks for flagging it!

**Reply by BastienLaby** [reply]
Hello Matt,

Unfortunately we actually don't have, at the time we speak, the resource to look in depth into Usd. We will choose a pipeline workaround by re-writing the node using the python API or a post-process event.

Thanks anyway for the detailed answer that I will keep aside for a potential future involvment.

Regards
